From 20260828104159.62399-2-yaseen@ghoul.dev Mon Sep 17 00:00:00 2001
From: Ahmed Yaseen <yaseen@xxxxxxxxx>
Date: Fri, 28 Aug 2026 06:43:33 -0400
Subject: [PATCH] HID: usbhid: skip interrupt IN polling for devices with no
 input reports

An input-less HID interface has no reports to receive through interrupt IN
polling. Avoid starting that polling path, since ASUS ROG N-Key keyboards
use a separate input-less interface for RGB/control operations and polling
it can make the sibling keyboard interface drop keypresses.

Reviewed-by: Denis Benato <denis.benato@xxxxxxxxx>
Signed-off-by: Ahmed Yaseen <yaseen@xxxxxxxxx>
---
 drivers/hid/usbhid/hid-core.c | 9 +++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/drivers/hid/usbhid/hid-core.c b/drivers/hid/usbhid/hid-core.c
index 96b0181cf819..d7cf2b56e117 100644
--- a/drivers/hid/usbhid/hid-core.c
+++ b/drivers/hid/usbhid/hid-core.c
@@ -688,7 +688,14 @@ static int usbhid_open(struct hid_device *hid)
 
 	set_bit(HID_OPENED, &usbhid->iofl);
 
-	if (hid->quirks & HID_QUIRK_ALWAYS_POLL) {
+	/*
+	 * ALWAYS_POLL devices are already polled from usbhid_start(), and a
+	 * device with no input reports has nothing to send on the interrupt
+	 * endpoint.  Avoid polling input-less interfaces because on ASUS ROG
+	 * N-Key keyboards it makes the sibling interface drop keypresses.
+	 */
+	if ((hid->quirks & HID_QUIRK_ALWAYS_POLL) ||
+	    list_empty(&hid->report_enum[HID_INPUT_REPORT].report_list)) {
 		res = 0;
 		goto Done;
 	}
-- 
2.55.0
